File: src/my_project/counterpoint/measure_search/rythmn_applyer.py
from my_project.counterpoint.measure_search.measure_step_sequence import (
    AbstractMeasureStepSequence,
    MeasureStepSequence,
)
from my_project.counterpoint.model import MeasureRythmnPattern, NoteAnnotation, ToneType
from my_project.model import Duration, IntervalStep, Measure, Melody, Note, Offset
from my_project.util import sliding


def try_apply_rythmn(
    step_sequence: MeasureStepSequence, rythmn_pattern: MeasureRythmnPattern
) -> Melody[Note[IntervalStep | None, NoteAnnotation]] | None:
    """
    MeasureStepSequence に対して MeasureRythmnPattern を適用できるかを判断する。
    可能なら適用した Melody を返し、不可能なら None を返す。
    """

    if step_sequence.num_notes_in_measure() != rythmn_pattern.measure_rythmn().num_durations():
        return None
    if step_sequence.is_tied_to_next_measure_required() != rythmn_pattern.measure_rythmn().is_next_tied:
        return None

    melody: Melody[Note[IntervalStep | None, NoteAnnotation]] = apply_rythmn(step_sequence, rythmn_pattern).melody

    # 次にタイで繋ぐ必要がある場合、最後の音は3拍目の2分音符で、和声音である必要がある。
    if step_sequence.is_tied_to_next_measure_required():
        offset_notes = melody.offset_notes()
        note_at_beat3 = offset_notes.get(Offset.idx_1(3))
        if (
            not note_at_beat3
            or note_at_beat3.duration != Duration.of(2)
            or note_at_beat3 is not melody.notes[-1]
            or note_at_beat3.attribute.tone_type
            != ToneType.HARMONIC_TONE  # 和声音である必要があるというのはここでやることか？
        ):
            return None

    # 掛留音
    if melody.notes[0].attribute.tone_type == ToneType.SUSPENDED_TONE:
        # 掛留音から始まる場合、リズムは前にタイが付いていないといけない
        if not rythmn_pattern.measure_rythmn().is_previous_tied:
            return None

        # 掛留音は3拍目に解決しなければならない
        offset3_note = melody.offset_notes().get(Offset.idx_1(3))
        if not offset3_note:
            return None
        if not offset3_note.attribute.tone_type == ToneType.HARMONIC_TONE:
            return None

        # 3拍目より前に解決してはいけない
        for offset, note in melody.offset_notes().items():
            if Offset.of(0) < offset < Offset.idx_1(3):
                if note.attribute.tone_type == ToneType.HARMONIC_TONE:
                    return None

    # 八分音符の周りに非順次進行を含めない
    # 次の小節の音も考慮するため、リストにダミーのノートを追加してループする
    next_note_dummy: Note[IntervalStep | None, NoteAnnotation] = Note(
        value=step_sequence.next_measure_step,
        duration=Duration.of(1),
        attribute=NoteAnnotation(is_tied_start=False, tone_type=ToneType.HARMONIC_TONE),
    )
    notes_to_check: list[Note[IntervalStep | None, NoteAnnotation]] = [*list(melody.notes), next_note_dummy]

    for current_note, next_note in sliding(notes_to_check, 2):
        if current_note.value is not None and current_note.duration == Duration.of(1, 2):
            if next_note is not None and next_note.value is not None:
                if (current_note.value - next_note.value).abs() >= IntervalStep.idx_1(3):
                    return None

    return melody


def apply_rythmn[T](
    step_sequence: AbstractMeasureStepSequence[T], rythmn_pattern: MeasureRythmnPattern
) -> Measure[Note[T | None, NoteAnnotation]]:
    """
    任意の AbstractMeasureStepSequence に対して MeasureRythmnPattern を適用した Melody を返す。
    ここではリズムの検証は行われない。事前に検証済みの内容に関して利用すること。
    """
    rythmn = rythmn_pattern.measure_rythmn()
    durations = rythmn.durations

    notes: list[Note[T | None, NoteAnnotation]] = []
    if rythmn.init_rest_duration > Duration.of(0):
        notes.append(
            Note(
                value=None,
                duration=rythmn.init_rest_duration,
                attribute=NoteAnnotation(is_tied_start=False, tone_type=ToneType.HARMONIC_TONE),
            )
        )

    for i, (note, duration) in enumerate(zip(step_sequence.measure.notes, durations)):
        is_last_note = i == step_sequence.num_notes_in_measure() - 1
        notes.append(
            Note(
                value=note.value,
                duration=duration,
                attribute=NoteAnnotation(
                    is_tied_start=step_sequence.is_tied_to_next_measure_required() and is_last_note,
                    tone_type=note.attribute,
                ),
            )
        )
    return Measure.of(*notes)

File: backend22/src/main/scala/composer/counterpoint/search/RythmnApplyer.scala
package composer.counterpoint.search

import composer.counterpoint.model.{MeasureRythmnPattern, NoteAnnotation, ToneType}
import model.containers.{Melody, Note}
import model.elements.{Duration, Offset, Part}
import model.elements.Interval.IntervalStep

object RythmnApplyer {

  case class AppliedIntervalStep(value: Option[IntervalStep], meta: NoteAnnotation)

  type AppliedNote   = Note[AppliedIntervalStep]
  type AppliedMelody = Melody[AppliedIntervalStep, AppliedNote]

  /** MeasureStepSequence に対して MeasureRythmnPattern を適用できるかを判断する。
    * 可能なら適用した Melody を返し、不可能なら None を返す。
    */
  def tryApplyRythmn(
      stepSequence: MeasureStepSequence,
      rythmnPattern: MeasureRythmnPattern,
  ): Option[AppliedMelody] = {
    val rythmn = rythmnPattern.measureRythmn

    if (stepSequence.numNotesInMeasure != rythmn.numDurations) return None
    if (stepSequence.isTiedToNextMeasureRequired != rythmn.isNextTied) return None

    val melody      = applyRythmn(stepSequence, rythmnPattern)
    val offsetNotes = calculateOffsetNotes(melody)

    // Next tie check
    if (stepSequence.isTiedToNextMeasureRequired) {
      val noteAtBeat3 = offsetNotes.get(Offset.idx1(3))
      val isValid     = noteAtBeat3.exists { n =>
        n.duration == Duration.of(2) &&
        n == melody.elems.last &&
        n.value.meta.toneType == ToneType.HARMONIC_TONE
      }
      if (!isValid) return None
    }

    // Suspension check
    if (melody.elems.head.value.meta.toneType == ToneType.SUSPENDED_TONE) {
      if (!rythmn.isPreviousTied) return None

      val offset3Note = offsetNotes.get(Offset.idx1(3))
      if (offset3Note.isEmpty) return None
      if (offset3Note.get.value.meta.toneType != ToneType.HARMONIC_TONE) return None

      // Should not resolve before beat 3
      val resolvedTooEarly = offsetNotes.exists { case (offset, note) =>
        offset > Offset.of(0) && offset < Offset.idx1(3) && note.value.meta.toneType == ToneType.HARMONIC_TONE
      }
      if (resolvedTooEarly) return None
    }

    // 8th note non-stepwise motion check
    // Create dummy next note
    val nextNoteDummy = Note(
      AppliedIntervalStep(
        Option(stepSequence.nextMeasureStep),
        NoteAnnotation(isTiedStart = false, ToneType.HARMONIC_TONE),
      ),
      Duration.of(1),
      Part.Root,
    )
    val notesToCheck = melody.elems :+ nextNoteDummy

    val pairs        = notesToCheck.zip(notesToCheck.tail)
    val hasViolation = pairs.exists { case (currentNote, nextNote) =>
      if (currentNote.value.value.isDefined && currentNote.duration == Duration.of(1, 2)) {
        if (nextNote.value.value.isDefined) {
          val diff = (currentNote.value.value.get - nextNote.value.value.get).abs
          diff >= IntervalStep.idx_1(3)
        } else false
      } else false
    }

    if (hasViolation) None else Some(melody)
  }

  /** 任意の AbstractMeasureStepSequence に対して MeasureRythmnPattern を適用した Melody を返す。
    * ここではリズムの検証は行われない。事前に検証済みの内容に関して利用すること。
    */
  def applyRythmn(
      stepSequence: AbstractMeasureStepSequence[IntervalStep],
      rythmnPattern: MeasureRythmnPattern,
  ): AppliedMelody = {
    val rythmn    = rythmnPattern.measureRythmn
    val durations = rythmn.durations

    var notes: List[AppliedNote] = Nil

    if (rythmn.initRestDuration > Duration.of(0)) {
      notes ::= Note(
        AppliedIntervalStep(
          None,
          NoteAnnotation(isTiedStart = false, ToneType.HARMONIC_TONE),
        ),
        rythmn.initRestDuration,
        Part.Root,
      )
    }

    val seqNotes = stepSequence.measureNotes
    for (i <- seqNotes.indices) {
      val note       = seqNotes(i)
      val duration   = durations(i)
      val isLastNote = i == stepSequence.numNotesInMeasure - 1

      notes ::= Note(
        AppliedIntervalStep(
          Some(note.value.value),
          NoteAnnotation(
            isTiedStart = stepSequence.isTiedToNextMeasureRequired && isLastNote,
            toneType = note.value.meta,
          ),
        ),
        duration,
        Part.Root,
      )
    }

    Melody(notes.reverse)
  }

  def calculateOffsetNotes(melody: AppliedMelody): Map[Offset, AppliedNote] = {
    var curr = Offset.of(0)
    melody.elems.map { note =>
      val entry = curr -> note
      curr = curr + note.duration
      entry
    }.toMap
  }
}
